---
title: "Level and plasticity divergence in a common trans-regulatory environment"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r package-loading}
source("functions_for_figure_scripts.R")
library(ggExtra)
load("data_files/FinalDataframe3Disp.RData")
load("data_files/Cleaned_Counts.RData")
load("data_files/Cleaned_Counts_Allele.RData")

cor_thresh <- 0.8
eff_thresh <- log2(1.5)
# extreme cor and log2 fold change values to extract the most
# cis plasticity divergers and most trans level divergers for
# avg expression figures
cis_plast_thresh <- -0.25 
trans_lev_thresh <- log2(2)
```

## Log2 fold changes between parents and hybrid are correlated

```{r level-explained, warning=FALSE}
# R squared for effect sizes between parents and hybrid for all genes
mod_full <- lm(effect_size_allele ~ effect_size_species, data = finaldf)
summary(mod_full)
# R squared genes with consistent effect size and diverged level in at least one environment
levdiv_genes <-  filter(finaldf, level == "diverged") |> 
  select(gene_name) |> pull() |> unique()
plotdf <- finaldf |> # filter(gene_name %in% levdiv_genes) |> 
  group_by(gene_name) # |> summarise(n_lev_dir = length(sign(effect_size_species))) |> 
  #filter(n_lev_dir == 1)
mod <- lm(effect_size_allele ~ effect_size_species, data = filter(finaldf, gene_name %in% levdiv_genes))
summary(mod) # more correlated for genes with the most consistent level divergence across envrironments

# looping through each experiment
library(ggExtra)
plotlist <- vector(mode = "list", length = length(ExperimentNames))
names(plotlist) <- ExperimentNames
for (e in ExperimentNames) {
  e_color <- colordf[colordf$limits == e,]$type
  plotdf <- finaldf |> filter(experiment == e) # & level == "diverged")
  mod_e <- lm(effect_size_allele ~ effect_size_species, data = plotdf)
  plotlist[[e]] <- ggMarginal(ggplot(plotdf, aes(x = effect_size_species,
                                 y = effect_size_allele)) +
               geom_vline(xintercept = 0, alpha = 0.5) +
               geom_hline(yintercept = 0, alpha = 0.5) +
               geom_abline(slope = 1, intercept = 0, alpha = 0.5) +
               geom_point(aes(color = level == "diverged"), alpha = 0.25, shape = 19) +
               # geom_text(x = -2, y = 4.5, 
               #           label = paste0("R^2 = ", round(summary(mod_e)$r.squared,
               #                                          digits = 3))) +
               scale_color_discrete(limits = c(TRUE, FALSE),
                                                     type = c("black", e_color)) +
               xlim(c(-5, 5)) +
               ylim(c(-5, 5)) +
               xlab("") +
               ylab("") +
               ggtitle(paste(LongExperimentNames[which(ExperimentNames == e)],
                       paste0(",\nR^2 = ", round(summary(mod_e)$r.squared,
                                                digits = 3)))) +
               theme_classic() +
               theme(legend.position = "none"),
             groupColour = TRUE, groupFill = TRUE)
}
```

Outputting plots to PDF:

```{r level-explained-pdf}
pdf("paper_figures/CisTrans/l2fc_eachEnvironment.pdf",
    width = 2.5, height = 16)
ggarrange(plotlist = plotlist, nrow = 6, ncol = 1)
dev.off()
ggarrange(plotlist = plotlist, nrow = 3, ncol = 2)
```

## Less of the plasticity divergence observed between parents is explained by the hybrid

```{r plasticity-unexplained, warning=FALSE}
# R squared for effect sizes between parents and hybrid for all genes
mod_full <- lm(cor_hybrid ~ cor_parents, data = finaldf)
summary(mod_full)
plotdf <- finaldf

# looping through each experiment
library(ggExtra)
plotlist <- vector(mode = "list", length = length(ExperimentNames))
names(plotlist) <- ExperimentNames
for (e in ExperimentNames) {
  e_color <- colordf[colordf$limits == e,]$type
  plotdf <- finaldf |> filter(experiment == e)
  mod_e <- lm(cor_hybrid ~ cor_parents, data = plotdf)
  plotlist[[e]] <- ggMarginal(ggplot(plotdf, aes(x = cor_parents,
                                                 y = cor_hybrid)) +
                                geom_vline(xintercept = 0, alpha = 0.5) +
                                geom_hline(yintercept = 0, alpha = 0.5) +
                                geom_abline(slope = 1, intercept = 0, alpha = 0.5) +
                                geom_point(aes(color = plasticity == "diverged"), alpha = 0.25, shape = 19) +
                                # geom_text(x = -0.5, y = 0.8, 
                                #           label = paste0("R^2 = ", round(summary(mod_e)$r.squared,
                                #                                          digits = 3))) +
                                scale_color_discrete(limits = c(TRUE, FALSE),
                                                     type = c("black", e_color)) +
               xlim(c(-1, 1)) +
               ylim(c(-1, 1)) +
               xlab("") +
               ylab("") +
               ggtitle(paste(LongExperimentNames[which(ExperimentNames == e)],
                       paste0(",\nR^2 = ", round(summary(mod_e)$r.squared,
                                                digits = 3)))) +
               theme_classic() +
               theme(legend.position = "none"),
             groupColour = TRUE, groupFill = TRUE)
}
```
```{r plasticity-unexplained-pdf}
pdf("paper_figures/CisTrans/cor_eachEnvironment.pdf",
    width = 2.5, height = 16)
ggarrange(plotlist = plotlist, nrow = 6, ncol = 1)
dev.off()
ggarrange(plotlist = plotlist, nrow = 3, ncol = 2)
```

## Hybrids have more correlated plasticity between alleles than parental orthologs

```{r plasticity-cor}
#### Distribution of parent/hybrid expression correlation ####
plotlist <- vector(mode = "list", length = length(ExperimentNames))
names(plotlist) <- ExperimentNames
# density plots of parental vs hybrid cor in each environment
for (e in ExperimentNames) {
  plotdf <- finaldf |> filter(experiment == e & plasticity == "diverged") |> 
    select(gene_name, experiment, cor_parents, cor_hybrid) |> 
    pivot_longer(cols = c("cor_parents", "cor_hybrid"), 
                 names_to = "parent_or_hybrid",
                 values_to = "cor")
  plotlist[[e]] <- ggplot(plotdf, aes(x = cor)) + 
    geom_density(data = filter(plotdf, parent_or_hybrid == "cor_parents"), 
                 aes(fill = experiment), alpha = 1, color = "white") +
    geom_density(data = filter(plotdf, parent_or_hybrid == "cor_hybrid"), 
                 aes(fill = experiment), alpha = 0.5, color = "black") +
    scale_fill_discrete(limits = colordf[colordf$scheme == "experiment",]$limits,
                        type = colordf[colordf$scheme == "experiment",]$type) +
    xlim(c(-1, 1)) +
    xlab("") +
    ylab("") +
    ggtitle(LongExperimentNames[which(ExperimentNames == e)]) +
    theme_classic() +
    theme(legend.position = "none")
}
p_null <- ggplot(plotdf) + theme_void()
pdf("paper_figures/CisTrans/cor_density_eachEnvironment.pdf",
    width = 8, height = 4)
ggarrange(plotlist[[1]], plotlist[[2]],
          p_null, p_null, plotlist[[3]], plotlist[[4]],
          plotlist[[5]], plotlist[[6]], nrow = 2, ncol = 4)
dev.off()
ggarrange(plotlist[[1]], plotlist[[2]],
          p_null, p_null, plotlist[[3]], plotlist[[4]],
          plotlist[[5]], plotlist[[6]], nrow = 2, ncol = 4)
```

## Level divergence tends to be in cis

What better way to show this than that looking at the small fraction of genes that are diverging in level in trans the most---that are diverging in level in the parents and have the smallest log2 fold change between hybrid alleles.

```{r level-pie-chart}
# pct genes in "transest level" group for figure
plotdf <- finaldf |> 
  filter(level == "diverged") |> 
  mutate(is_trans_upcer = (effect_size_species > trans_lev_thresh) &
           (abs(effect_size_allele) < eff_thresh),
         is_trans_uppar = (effect_size_species < -trans_lev_thresh) &
           (abs(effect_size_allele) < eff_thresh))
plotdf |> mutate(is_trans = is_trans_upcer | is_trans_uppar) |> 
  group_by(experiment) |> 
  summarise(pct_trans = sum(is_trans)/n(),
            pct_rest = sum(!is_trans)/n())
pctdf <- plotdf |> mutate(is_trans = is_trans_upcer | is_trans_uppar) |> 
  group_by(experiment) |> 
  summarise(trans = sum(is_trans)/n(),
            cis = sum(!is_trans)/n()) |> 
  pivot_longer(cols = c("cis", "trans"), names_to = "mode",
               values_to = "percent")
p <- ggplot(pctdf, aes(x="", y=percent, fill=mode)) +
  scale_fill_discrete(limits = c("cis", "trans"),
                      type = c("salmon", "aquamarine")) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0) +
  theme_classic() +
  theme(axis.ticks = element_blank(),
        axis.text = element_blank(),
        legend.position = "bottom") +
  xlab("") +
  ylab("") +
  facet_wrap(~factor(experiment, levels = ExperimentNames,
                     labels = LongExperimentNames), nrow = 2, ncol = 3)
pdf("paper_figures/CisTrans/level_pcts.pdf",
    width = 8, height = 4)
p
dev.off()
p
```

Average expression of the cis and trans portion of level divergers, Diauxic Shift as example:

```{r level-avg-expr}
plotdf <- finaldf |> 
  filter(level == "diverged") |> 
  mutate(is_trans_upcer = (effect_size_species > trans_lev_thresh) &
           (abs(effect_size_allele) < eff_thresh),
         is_trans_uppar = (effect_size_species < -trans_lev_thresh) &
           (abs(effect_size_allele) < eff_thresh))
plotdf |> mutate(is_trans = is_trans_upcer | is_trans_uppar) |> 
  group_by(experiment) |> 
  summarise(pct_trans = sum(is_trans)/n(),
            pct_rest = sum(!is_trans)/n())
# upcer, trans
gene_idxs <- filter(plotdf, experiment == "HAP4" & cer == 1 & 
                      par == 1 & is_trans_upcer) |> 
  select(gene_name) |> pull()
p <- annotate_figure(plotGenes(gene_idxs, .quartet = TRUE,
                               .experiment_name = "HAP4"), 
                     top = paste(length(gene_idxs), "genes"))
p
pdf("paper_figures/CisTrans/level_trans_HAP411_upcer.pdf", width = 3, height = 3)
p
dev.off()
  
# upcer, cis
gene_idxs <- filter(plotdf, experiment == "HAP4" & cer == 1 & 
                      par == 1 & !is_trans_upcer & !is_trans_uppar &
                        effect_size_species > 0) |> 
  select(gene_name) |> pull()
p <- annotate_figure(plotGenes(gene_idxs, .quartet = TRUE,
                               .experiment_name = "HAP4"), 
                     top = paste(length(gene_idxs), "genes"))
p
pdf("paper_figures/CisTrans/level_cis_HAP411_upcer.pdf", width = 3, height = 3)
p
dev.off()
```

## Plasticity divergence tends to be in trans

```{r plasticity-pie-chart}
plotdf <- finaldf |> 
  filter(plasticity == "diverged") |> 
  mutate(is_cis = cor_hybrid < cis_plast_thresh)
plotdf |> group_by(experiment) |> 
  summarise(cis = sum(is_cis, na.rm = TRUE)/n(),
            trans = sum(!is_cis, na.rm = TRUE)/n())
pctdf <- plotdf |> group_by(experiment) |> 
  summarise(cis = sum(is_cis, na.rm = TRUE)/n(),
            trans = sum(!is_cis, na.rm = TRUE)/n()) |> 
  pivot_longer(cols = c("cis", "trans"), names_to = "mode",
               values_to = "percent")
p <- ggplot(pctdf, aes(x="", y=percent, fill=mode)) +
  scale_fill_discrete(limits = c("cis", "trans"),
                      type = c("salmon", "aquamarine")) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0) +
  theme_classic() +
  theme(axis.ticks = element_blank(),
        axis.text = element_blank(),
        legend.position = "bottom") +
  xlab("") +
  ylab("") +
  facet_wrap(~factor(experiment, levels = ExperimentNames,
                     labels = LongExperimentNames), nrow = 2, ncol = 3)
p
pdf("paper_figures/CisTrans/plasticity_pcts.pdf",
    width = 8, height = 4)
p
dev.off()
```
Average expression of Diauxic Shift example:

```{r plasticity-avg-expr}
plotdf <- finaldf |> 
  filter(plasticity == "diverged") |> 
  mutate(is_cis = cor_hybrid < cis_plast_thresh)
plotdf |> group_by(experiment) |> 
  summarise(cis = sum(is_cis, na.rm = TRUE)/n(),
            trans = sum(!is_cis, na.rm = TRUE)/n())
# cis
gene_idxs <- filter(plotdf, experiment == "HAP4" & cer == 2 & 
                        par == 1 & is_cis) |> 
    select(gene_name) |> pull()
p <- annotate_figure(plotGenes(gene_idxs, .quartet = TRUE,
                                    .experiment_name = "HAP4"), 
                          top = paste(length(gene_idxs), "genes"))
p
pdf("paper_figures/CisTrans/plasticity_cis_HAP421.pdf",
        width = 3, height = 3)
p
dev.off()

# trans
gene_idxs <- filter(plotdf, experiment == "HAP4" & cer == 2 & 
                        par == 1 & !is_cis) |> 
    select(gene_name) |> pull()
p <- annotate_figure(plotGenes(gene_idxs, .quartet = TRUE,
                                    .experiment_name = "HAP4"), 
                          top = paste(length(gene_idxs), "genes"))
p
pdf("paper_figures/CisTrans/plasticity_trans_HAP421.pdf",
        width = 3, height = 3)
p
dev.off()
```


