<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Data cleaning for figure scripts</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Plasticity</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    1-3: Data Cleaning
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="clean_data.html">1. Import data and QC</a>
    </li>
    <li>
      <a href="data_for_analysis_scripts.html">2. Data cleaning</a>
    </li>
    <li>
      <a href="functions_for_figure_scripts.html">3. Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    4-6: Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="clustering.html">4. Clustering</a>
    </li>
    <li>
      <a href="single_gene_models.html">5. Generalized linear models</a>
    </li>
    <li>
      <a href="data_for_figure_scripts.html">6. Final dataframe</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    7: Single-environment figures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Fig_DiauxicShift.html">Fig 5: Diauxic Shift</a>
    </li>
    <li>
      <a href="Fig_HUShock.html">Fig S3: HU Shock</a>
    </li>
    <li>
      <a href="Fig_LowNitrogen.html">Fig S3: Low Nitrogen</a>
    </li>
    <li>
      <a href="Fig_LowPhosphate.html">Fig S3: Low Phosphate</a>
    </li>
    <li>
      <a href="Fig_HeatStress.html">Fig S3: Heat Stress</a>
    </li>
    <li>
      <a href="Fig_ColdStress.html">Fig S3: Cold Stress</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    8-10: Multi-environment figures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Fig_gene_ontology.html">8. Table 1: Gene Ontology</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Data cleaning for figure scripts</h1>

</div>


<p>Final data cleaning steps that import level and plasticity divergence
information from DESeq2 and Heiarchical clustering. The end result of
this script is a dataframe, finaldf, that is used in all the figure
scripts (also as Supplementary Table 1)</p>
<pre class="r"><code>sapply(c(&quot;dplyr&quot;, &quot;purrr&quot;, &quot;tidyr&quot;, &quot;ggpubr&quot;, &quot;readr&quot;,
         &quot;data.table&quot;, &quot;ggplot2&quot;, &quot;data.table&quot;, 
         &quot;matrixStats&quot;), require, character.only=TRUE)</code></pre>
<pre><code>##       dplyr       purrr       tidyr      ggpubr       readr 
##        TRUE        TRUE        TRUE        TRUE        TRUE 
##  data.table     ggplot2  data.table matrixStats 
##        TRUE        TRUE        TRUE        TRUE</code></pre>
<pre class="r"><code>source(&quot;functions_for_figure_scripts.R&quot;)</code></pre>
<p>Defining our color schemes and category names to be used throughout
the paper figures</p>
<pre class="r"><code># Environment Names, so I only have to change them in one place
ExperimentNames &lt;- c(&quot;HAP4&quot;, &quot;CC&quot;, &quot;LowN&quot;, &quot;LowPi&quot;, &quot;Heat&quot;, &quot;Cold&quot;)
LongExperimentNames &lt;- c(&quot;Diauxic Shift&quot;, &quot;Hydroxyurea Shock&quot;, &quot;Low Nitrogen&quot;, &quot;Low Phosphate&quot;, &quot;Heat Stress&quot;, &quot;Cold Stress&quot;)

# colors used throughout paper
group4_colordf &lt;- tibble(type = c(&quot;grey80&quot;, &quot;gold&quot;, &quot;grey80&quot;, &quot;gold&quot;, &quot;grey20&quot;),
                         labels = c(&quot;conserved level \nand plasticity&quot;, 
                                    &quot;conserved level, \ndiverged plasticity&quot;, 
                                    &quot;diverged level, \nconserved plasticity&quot;, 
                                    &quot;diverged level \nand plasticity&quot;,
                                    &quot;lowly expressed&quot;),
                         limits = c(&quot;conserved level and plasticity&quot;, 
                                    &quot;conserved level, diverged plasticity&quot;, 
                                    &quot;diverged level, conserved plasticity&quot;, 
                                    &quot;diverged level and plasticity&quot;,
                                    &quot;lowly expressed&quot;),
                         pattern = c(&quot;none&quot;, &quot;none&quot;,
                                     &quot;stripe&quot;, &quot;stripe&quot;,
                                     &quot;none&quot;),
                         scheme = &quot;group4&quot;)
experiment_colordf &lt;- tibble(type = c(&quot;grey&quot;, &quot;yellow&quot;, &quot;green&quot;, &quot;purple&quot;, &quot;red&quot;, &quot;blue&quot;),
                                               labels = LongExperimentNames,
                                               limits = ExperimentNames,
                                               scheme = &quot;experiment&quot;)
plasticity_colordf &lt;- tibble(type = c(&quot;grey80&quot;, &quot;grey60&quot;, &quot;orange1&quot;, &quot;blue2&quot;, &quot;purple&quot;),
                           labels = c(&quot;conserved plastic&quot;, &quot;conserved static&quot;, &quot;Scer-unique plasticity&quot;, &quot;Spar-unique plasticity&quot;, &quot;plasticity reversal&quot;),
                           limits = c(&quot;conserved plastic&quot;, &quot;conserved static&quot;, &quot;Scer-unique&quot;, &quot;Spar-unique&quot;, &quot;reversal&quot;),
                           scheme = &quot;plasticity&quot;)
colordf &lt;- bind_rows(group4_colordf, experiment_colordf, plasticity_colordf)</code></pre>
<div id="creating-final-dataframe" class="section level2">
<h2>Creating final dataframe</h2>
<p>Loading heiarchical clustering and DESeq2 model outputs:</p>
<pre class="r"><code>load(&quot;data_files/CorrelationClustering3Disp.RData&quot;)
load(&quot;data_files/single_gene_models.RData&quot;)
load(&quot;data_files/Cleaned_Count_Data.RData&quot;)
load(&quot;data_files/Cleaned_Counts.RData&quot;)
load(&quot;data_files/Cleaned_Counts_Allele.RData&quot;)
p_thresh &lt;- 0.05
eff_thresh &lt;- log2(1.5) # gene needs to be 1.5x higher in one species than the other to be considered DE
eff_thresh0 &lt;- 0
eff_thresh2 &lt;- log2(2)</code></pre>
<p>Benjamini-Hochberg FDR correction:</p>
<pre class="r"><code># Benjamini-Hochberg FDR correction of glm p-values
spaldf$padj &lt;- p.adjust(spaldf$pvalue, method = &quot;BH&quot;)
sum(spaldf$pvalue &lt; p_thresh)</code></pre>
<pre><code>## [1] 25607</code></pre>
<pre class="r"><code>sum(spaldf$padj &lt; p_thresh)</code></pre>
<pre><code>## [1] 22628</code></pre>
<p>Creating mutually exclusive sets of genes based on level and
plasticity divergence in parents each gene is a combination of its level
category and its plasticity category ## level categories: 1) conserved:
parent padj &gt;= p_thresh 2) diverged: parent padj &lt; p_thresh, not
taking into account hybrid yet ## plasticity categories (just looking at
parents): 1) conserved: 11, 22, 00 2) diverged: 12, 21, 01, 10, 02,
20</p>
<pre class="r"><code>finaldf &lt;- pivot_wider(spaldf, id_cols = c(&quot;gene_name&quot;, &quot;experiment&quot;), 
                       values_from = c(&quot;effect_size&quot;, &quot;padj&quot;),
                       names_from = &quot;coefficient&quot;) |&gt; 
  drop_na() |&gt; # 1 gene missing from species, YMR107W
  inner_join(y =  clusterdf, 
             by = join_by(&quot;gene_name&quot;==&quot;gene_ID&quot;,
                          &quot;experiment&quot;))

### Removing hybrid par allele-deleted genes from CC experiment (where the deletion was present)
omit_list &lt;- c(&quot;YLR078C&quot;, &quot;YLR077W&quot;, &quot;YLR074C&quot;, &quot;YLR072W&quot;, &quot;YLR075W&quot;, &quot;YLR073C&quot;, # large hybrid CC paradoxus haplotype deletion
               &quot;YNL247W&quot;, &quot;YNL244C&quot;)
finaldf |&gt; filter(experiment == &quot;CC&quot; &amp; gene_name %in% omit_list)</code></pre>
<table class="huxtable" data-quarto-disable-processing="true" style="border-collapse: collapse; border: 0px; margin-bottom: 2em; margin-top: 2em; ; margin-left: auto; margin-right: auto;  " id="tab:finaldf">
<col><col><col><col><col><col><col><col><tr>
<th style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">gene_name</th><th style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">experiment</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">effect_size_allele</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">effect_size_species</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">padj_allele</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">padj_species</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">cer</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0.4pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">par</th></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">YLR072W</td><td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">8.76</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">-0.0546</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1.19e-27&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0.827&nbsp;&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">YLR073C</td><td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">9.16</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">0.424&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">2.6e-16&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">0.382&nbsp;&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">1</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">1</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">YLR074C</td><td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">9.85</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">-0.0309</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">4.67e-40&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0.897&nbsp;&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">2</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">YLR075W</td><td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">8.26</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">-0.469&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">5.33e-139</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">5.87e-09</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">2</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">2</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">YLR077W</td><td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">42.4&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0.93&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1e-06&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">YLR078C</td><td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">42.9&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">0.469&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">0.0424&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">1</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">1</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">YNL247W</td><td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">12&nbsp;&nbsp;&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0.429&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">4.31e-30&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0.000508</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">2</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">2</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">YNL244C</td><td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">8.7&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">-0.717&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">1.81e-186</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">3.58e-17</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">2</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">2</td></tr>
</table>

<pre class="r"><code>dim(finaldf)</code></pre>
<pre><code>## [1] 25264     8</code></pre>
<pre class="r"><code>finaldf &lt;- finaldf |&gt; filter(!(experiment == &quot;CC&quot; &amp; gene_name %in% omit_list))
dim(finaldf)</code></pre>
<pre><code>## [1] 25256     8</code></pre>
<p>Adding plasticity column. If cluster in Scer is the same as cluster
in Spar, plasticity is conserved. If clusters are different, plasticity
is diverged.</p>
<pre class="r"><code>finaldf$plasticity &lt;- map2(finaldf$cer, finaldf$par, \(x, y) {
  if (x == y) {
    return(&quot;conserved&quot;)
  }
  if (x != y) {
    return(&quot;diverged&quot;)
  }
}) |&gt; unlist()</code></pre>
<p>Adding level column. If the DEseq2-estimated log2 fold change when
you change from an Spar count to an Scer count (Spar is reference) is
greater (in magnitude) than the log2 fold change threshold (eff_thresh)
and the BH-adjusted p-value is less than 0.05, the gene is considered
diverged in level. Using three different log2 fold change thresholds to
compare the effect in a Supplemental figure.</p>
<pre class="r"><code>finaldf$level &lt;- map(c(1:nrow(finaldf)), \(i) {
  x &lt;- finaldf$padj_species[i] |&gt; as.numeric()
  y &lt;- finaldf$effect_size_species[i] |&gt; as.numeric()
  output &lt;- if_else(x &lt; p_thresh &amp; abs(y) &gt; eff_thresh,
                    true = &quot;diverged&quot;,
                    false = &quot;conserved&quot;)
  return(output)
}) |&gt; unlist()
finaldf$level0 &lt;- map(c(1:nrow(finaldf)), \(i) {
  x &lt;- finaldf$padj_species[i] |&gt; as.numeric()
  y &lt;- finaldf$effect_size_species[i] |&gt; as.numeric()
  output &lt;- if_else(x &lt; p_thresh &amp; abs(y) &gt; eff_thresh0,
                    true = &quot;diverged&quot;,
                    false = &quot;conserved&quot;)
  return(output)
}) |&gt; unlist()
finaldf$level2 &lt;- map(c(1:nrow(finaldf)), \(i) {
  x &lt;- finaldf$padj_species[i] |&gt; as.numeric()
  y &lt;- finaldf$effect_size_species[i] |&gt; as.numeric()
  output &lt;- if_else(x &lt; p_thresh &amp; abs(y) &gt; eff_thresh2,
                    true = &quot;diverged&quot;,
                    false = &quot;conserved&quot;)
  return(output)
}) |&gt; unlist()</code></pre>
<p>Tables of gene counts to report in main paper:</p>
<pre class="r"><code># gene counts to report
table(finaldf$plasticity, finaldf$level, useNA = &quot;always&quot;) # checking that all observations are in the 2x2 box</code></pre>
<pre><code>##            
##             conserved diverged  &lt;NA&gt;
##   conserved     10020     5436     0
##   diverged       6244     3556     0
##   &lt;NA&gt;              0        0     0</code></pre>
<pre class="r"><code>sum(table(finaldf$level, finaldf$plasticity))</code></pre>
<pre><code>## [1] 25256</code></pre>
<pre class="r"><code>nrow(finaldf) # mutually exclusive group categories</code></pre>
<pre><code>## [1] 25256</code></pre>
<pre class="r"><code>finaldf |&gt; filter(plasticity == &quot;diverged&quot;) |&gt; select(experiment) |&gt; table()</code></pre>
<pre><code>## experiment
##    CC  Cold  HAP4  Heat  LowN LowPi 
##  1766  1786  1455  1743  1305  1745</code></pre>
<pre class="r"><code>finaldf |&gt; filter(level == &quot;diverged&quot;) |&gt; select(experiment) |&gt; table()</code></pre>
<pre><code>## experiment
##    CC  Cold  HAP4  Heat  LowN LowPi 
##  1532  1563  1238  1415  1657  1587</code></pre>
<p>Tables of number of timepoints per experiment for paper figure:</p>
<pre class="r"><code>sample_info |&gt; select(experiment, time_point_str) |&gt; 
  unique() |&gt; 
  group_by(experiment) |&gt; 
  summarise(n = n())</code></pre>
<table class="huxtable" data-quarto-disable-processing="true" style="border-collapse: collapse; border: 0px; margin-bottom: 2em; margin-top: 2em; ; margin-left: auto; margin-right: auto;  " id="tab:paper-table-samples">
<col><col><tr>
<th style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">experiment</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0.4pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">n</th></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">8</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">Cold</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">4</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">HAP4</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">15</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">Heat</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">4</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">LowN</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">3</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">LowPi</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">28</td></tr>
</table>

<p>Calculating ortholog pair correlations for EnvironmentalPatterns
figure</p>
<pre class="r"><code># between hybrid alleles and for parent-hybrid allele of the same species (Hyc/Scer and Hyp/Spar)
getCorelation &lt;- function(.gene_name, .experiment, 
                          .cts1, .cts2, .info1, .info2) {
  common_cols &lt;- intersect(colnames(.cts1[,.info1$experiment == .experiment]), 
                           colnames(.cts2[,.info2$experiment == .experiment]))
  output &lt;- cor(as.numeric(.cts1[.gene_name, common_cols]),
                as.numeric(.cts2[.gene_name, common_cols]))
  return(as.numeric(output))
}
# tests for getCorelation
getCorelation(&quot;YGR192C&quot;, &quot;HAP4&quot;,
              .cts1 = collapsed$cer,
              .cts2 = collapsed_allele$cer,
              .info1 = info,
              .info2 = info_allele)</code></pre>
<pre><code>## [1] 0.9640884</code></pre>
<p>Adding correlation columns:</p>
<pre class="r"><code># adding correlation columns
finaldf$cor_hybrid &lt;- map2(finaldf$gene_name, finaldf$experiment, getCorelation,
                           .cts1 = collapsed_allele$cer,
                           .cts2 = collapsed_allele$par,
                           .info1 = info_allele,
                           .info2 = info_allele) |&gt; unlist()
finaldf$cor_parents &lt;- map2(finaldf$gene_name, finaldf$experiment, getCorelation,
                            .cts1 = collapsed$cer,
                            .cts2 = collapsed$par,
                            .info1 = info,
                            .info2 = info) |&gt; unlist()
finaldf$cor_scer &lt;- map2(finaldf$gene_name, finaldf$experiment, getCorelation,
                         .cts1 = collapsed_allele$cer,
                         .cts2 = collapsed$cer,
                         .info1 = info_allele,
                         .info2 = info) |&gt; unlist()
finaldf$cor_spar &lt;- map2(finaldf$gene_name, finaldf$experiment, getCorelation,
                         .cts1 = collapsed_allele$par,
                         .cts2 = collapsed$par,
                         .info1 = info_allele,
                         .info2 = info) |&gt; unlist()</code></pre>
<p>Calculating the mean and var of each gene’s expression across both
species, which is mainly relevant in a couple Supplemental/QC plots</p>
<p>mean:</p>
<pre class="r"><code>getMean &lt;- function(.gene_name, .experiment, 
                   .cts, .info) {
  output &lt;- rowMeans(.cts[.gene_name, .info$experiment == .experiment, drop = FALSE])
  return(as.numeric(output))
}
# tests for getMean
getMean(&quot;YGR192C&quot;, &quot;HAP4&quot;,
        .cts = counts,
        .info = sample_info)</code></pre>
<pre><code>## [1] 9339.933</code></pre>
<p>var:</p>
<pre class="r"><code>getVar &lt;- function(.gene_name, .experiment, 
                    .cts, .info) {
  output &lt;- rowVars(.cts[.gene_name, .info$experiment == .experiment, drop = FALSE])
  return(as.numeric(output))
}
# tests for getVar
getVar(&quot;YGR192C&quot;, &quot;HAP4&quot;,
        .cts = counts,
        .info = sample_info)</code></pre>
<pre><code>## [1] 67159939</code></pre>
<p>calculating for each gene (including both parent’s counts):</p>
<pre class="r"><code># adding mean&amp;var columns
finaldf$mean_parents &lt;- map2(finaldf$gene_name, finaldf$experiment, getMean,
                             .cts = counts,
                             .info = sample_info) |&gt; unlist()
finaldf$var_parents &lt;- map2(finaldf$gene_name, finaldf$experiment, getVar,
                             .cts = counts,
                             .info = sample_info) |&gt; unlist()</code></pre>
<p>Remember those biased genes we identified in the very first importing
count data script? These were genes that had a high proportion of counts
mapping to the wrong species allele. They become relevant here. We
re-label any gene that has been flagged as diverged in level as “biased”
instead of “diverged”</p>
<pre class="r"><code>finaldf &lt;- finaldf |&gt; mutate(bias = if_else(gene_name %in% cer_biased_genes,
                                           true = &quot;cer&quot;,
                                           false = if_else(gene_name %in% par_biased_genes,
                                                           true = &quot;par&quot;,
                                                           false = if_else(gene_name %in% both_biased_genes,
                                                                           true = &quot;both&quot;,
                                                                           false = &quot;none&quot;)))) 
plotdf &lt;- finaldf |&gt; 
  filter(level == &quot;diverged&quot;)
# effect of bias on level divergence
ggplot(plotdf, aes(x = effect_size_species)) +
  geom_density(aes(fill = bias), alpha = 0.5) </code></pre>
<p><img src="data_for_figure_scripts_files/figure-html/bias-1.png" width="672" /></p>
<p>Re-labeling these biased genes:</p>
<pre class="r"><code>finaldf$level &lt;- if_else(finaldf$bias == &quot;none&quot;,
                         true = finaldf$level,
                         false = &quot;biased&quot;)

# effect of bias on plasticity divergence
p_none &lt;- plotdf |&gt; filter(bias == &quot;none&quot;) |&gt; 
  select(cer, par) |&gt; table()
p_cer &lt;- plotdf |&gt; filter(bias == &quot;cer&quot;) |&gt; 
  select(cer, par) |&gt; table()
p_par &lt;- plotdf |&gt; filter(bias == &quot;par&quot;) |&gt; 
  select(cer, par) |&gt; table()

round(p_none/sum(p_none), digits = 2)</code></pre>
<pre><code>##    par
## cer    0    1    2
##   0 0.10 0.08 0.02
##   1 0.10 0.38 0.05
##   2 0.05 0.09 0.13</code></pre>
<pre class="r"><code>round(p_cer/sum(p_cer), digits = 2)</code></pre>
<pre><code>##    par
## cer    0    1    2
##   0 0.05 0.01 0.03
##   1 0.15 0.59 0.05
##   2 0.06 0.03 0.03</code></pre>
<pre class="r"><code>round(p_par/sum(p_par), digits = 2)</code></pre>
<pre><code>##    par
## cer    0    1    2
##   0 0.03 0.13 0.09
##   1 0.04 0.52 0.08
##   2 0.02 0.03 0.07</code></pre>
<pre class="r"><code># proportions are the same between the three groups</code></pre>
<p>Numbers for the workflow figure:</p>
<pre class="r"><code>### numbers for workflow figure
nrow(finaldf) # for workflow figure E</code></pre>
<pre><code>## [1] 25256</code></pre>
<pre class="r"><code>length(unique(finaldf$gene_name)) # total number of genes</code></pre>
<pre><code>## [1] 4863</code></pre>
<pre class="r"><code>n_per_experiment &lt;- table(finaldf$experiment)
n_per_experiment # number of genes used in each environment (number expressed &gt; 30cpm)</code></pre>
<pre><code>## 
##    CC  Cold  HAP4  Heat  LowN LowPi 
##  3952  4051  4286  4172  4412  4383</code></pre>
<pre class="r"><code># min filtered out for low expression:
max(n_per_experiment)</code></pre>
<pre><code>## [1] 4412</code></pre>
<pre class="r"><code>(nrow(genedf) - max(n_per_experiment))/nrow(genedf)</code></pre>
<pre><code>## [1] 0.09274111</code></pre>
<pre class="r"><code># max filtered out for low expression:
min(n_per_experiment)</code></pre>
<pre><code>## [1] 3952</code></pre>
<pre class="r"><code>(nrow(genedf) - min(n_per_experiment))/nrow(genedf)</code></pre>
<pre><code>## [1] 0.1873329</code></pre>
<pre class="r"><code># percent low variance per experiment:
finaldf |&gt; group_by(experiment) |&gt; 
  summarise(n_cer_lowvar = sum(cer == 0),
            n_par_lowvar = sum(par == 0),
            n_genes = n()*2) |&gt; 
  mutate(pct_low_var = (n_cer_lowvar + n_par_lowvar)/n_genes)</code></pre>
<table class="huxtable" data-quarto-disable-processing="true" style="border-collapse: collapse; border: 0px; margin-bottom: 2em; margin-top: 2em; ; margin-left: auto; margin-right: auto;  " id="tab:workflow-numbers">
<col><col><col><col><col><tr>
<th style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">experiment</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">n_cer_lowvar</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">n_par_lowvar</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">n_genes</th><th style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0.4pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: bold;">pct_low_var</th></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">CC</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">315</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">617</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">7.9e+03&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0.4pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0.118&nbsp;</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">Cold</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">1585</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">2148</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">8.1e+03&nbsp;</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">0.461&nbsp;</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">HAP4</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">971</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">850</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">8.57e+03</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0.212&nbsp;</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">Heat</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">833</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">1207</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">8.34e+03</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">0.244&nbsp;</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">LowN</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1300</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">1196</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">8.82e+03</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0pt 0pt;    padding: 6pt 6pt 6pt 6pt; background-color: rgb(242, 242, 242); font-weight: normal;">0.283&nbsp;</td></tr>
<tr>
<td style="vertical-align: top; text-align: left; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0.4pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">LowPi</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">372</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">253</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">8.77e+03</td><td style="vertical-align: top; text-align: right; white-space: normal; border-style: solid solid solid solid; border-width: 0pt 0.4pt 0.4pt 0pt;    padding: 6pt 6pt 6pt 6pt; font-weight: normal;">0.0713</td></tr>
</table>

<div
id="how-the-choice-of-effect-size-threshold-influence-level-divergence"
class="section level3">
<h3>How the choice of effect size threshold influence level
divergence</h3>
<p>This is slightly sacrilegious to put a figure plot in this
data-cleaning script. But it is easier than having a separate script
just for this figure. Here we compare how the three different effect
size thresholds affect the number of genes considered to be diverging in
expression level.</p>
<pre class="r"><code># for testing how eff_threshold affects level divergence
finaldf &lt;- finaldf |&gt; pivot_longer(cols = c(&quot;level0&quot;, &quot;level&quot;, &quot;level2&quot;),
                                   names_to = &quot;eff_threshold&quot;,
                                   values_to = &quot;level&quot;)

### Barplot of 4 divergence categories
finaldf$group4 &lt;- map2(finaldf$level, finaldf$plasticity, \(l, d) {
  if (l != &quot;diverged&quot; &amp; d == &quot;conserved&quot;) {
    return(&quot;conserved level and plasticity&quot;)
  }
  if (l != &quot;diverged&quot; &amp; d == &quot;diverged&quot;) {
    return(&quot;conserved level, diverged plasticity&quot;)
  }
  if (l == &quot;diverged&quot; &amp; d == &quot;conserved&quot;) {
    return(&quot;diverged level, conserved plasticity&quot;)
  }
  if (l == &quot;diverged&quot; &amp; d == &quot;diverged&quot;) {
    return(&quot;diverged level and plasticity&quot;)
  }
}) |&gt; unlist()

plotdf &lt;- finaldf
plotdf$eff_threshold &lt;- factor(plotdf$eff_threshold, levels = c(&quot;level0&quot;, &quot;level&quot;, &quot;level2&quot;),
                               labels = c(&quot;no threshold&quot;, &quot;1.5x&quot;, &quot;2x&quot;))
p &lt;- ggplot(mutate(plotdf, experiment = factor(experiment,
                                          levels = ExperimentNames)),
       aes(x = group4)) +
  geom_bar_pattern(aes(fill = group4, pattern = group4), pattern_fill = &quot;black&quot;) +
  geom_text(stat=&#39;count&#39;, aes(label = after_stat(count)), vjust=-1) +
  scale_fill_discrete(limits =  colordf[colordf$scheme == &quot;group4&quot;,]$limits,
                      type = colordf[colordf$scheme == &quot;group4&quot;,]$type) +
  scale_pattern_discrete(limits =  colordf[colordf$scheme == &quot;group4&quot;,]$limits,
                         choices = colordf[colordf$scheme == &quot;group4&quot;,]$pattern) +
  scale_x_discrete(limits = colordf[colordf$scheme == &quot;group4&quot;,]$limits,
                   breaks = colordf[colordf$scheme == &quot;group4&quot;,]$limits,
                   labels = colordf[colordf$scheme == &quot;group4&quot;,]$labels) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        legend.position = &quot;bottom&quot;) +
  xlab(&quot;&quot;) +
  ylim(c(0, 3500)) +
  facet_grid(eff_threshold ~ experiment) +
  guides(fill = guide_legend(nrow = 3, byrow=TRUE))
p</code></pre>
<p><img src="data_for_figure_scripts_files/figure-html/level-threshold-qc-1.png" width="672" /></p>
<pre class="r"><code>pdf(&quot;paper_figures/Supplement/bar.pdf&quot;,
    width = 11, height = 5)
p
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<p>Removing level thresholds that we don’t use, just keeping the 1.5
threshold</p>
<pre class="r"><code># removing effect size threshold we don&#39;t use
finaldf &lt;- filter(finaldf, eff_threshold == &quot;level&quot;) |&gt; 
  select(-eff_threshold)</code></pre>
</div>
<div id="group-codes" class="section level3">
<h3>Group codes</h3>
<p>I am 99% sure these group codes are no longer used in any of the
figure scripts, but keeping them in just in case. These incorporate all
the level and plasticiy divergence information from the other columns
into one column The idea was that these codes might effectively group
certain types of genes together more than the individual properties. As
we discovered in the course of the analyses of the paper, however, level
and plasticity divergence do not have any relationship with each other,
and the types of plasticity divergence observed tended to have more to
do with the environment than the particular genes that had that
divergence pattern.</p>
<pre class="r"><code>### longer group codes
# gene groups factor in a) whether level is conserved, b) whether dynamics are conserved,
# and c) if level or dynamics are diverged, which direction that is happening in (i.e. up in Scer, or 2 in Scer and 1 in Spar)
# group code format: type_direction_cluster(s)
# type: &lt;cons, lev, dyn, or levdyn&gt;
# direction (of level divergence): &lt;uppar, upcer, or empty&gt;
# clusters: &lt;0, 1, 2, 01, 02, etc.&gt;
finaldf$group &lt;- apply(finaldf, 1, \(x) {
  x_cer &lt;- x[&quot;cer&quot;] |&gt; as.numeric()
  x_par &lt;- x[&quot;par&quot;] |&gt; as.numeric()
  x_effect_size &lt;- x[&quot;effect_size_species&quot;] |&gt; as.numeric()
  x_level &lt;- x[&quot;level&quot;] |&gt; as.character()
  x_plasticity &lt;- x[&quot;plasticity&quot;] |&gt; as.character()
  if (x_level != &quot;diverged&quot; &amp; x_plasticity == &quot;conserved&quot;) {
    type_str &lt;- &quot;cons&quot;
  }
  if (x_level == &quot;diverged&quot; &amp; x_plasticity == &quot;conserved&quot;) {
    type_str &lt;- &quot;lev&quot;
  }
  if (x_level != &quot;diverged&quot; &amp; x_plasticity == &quot;diverged&quot;) {
    type_str &lt;- &quot;dyn&quot;
  }
  if (x_level == &quot;diverged&quot; &amp; x_plasticity == &quot;diverged&quot;) {
    type_str &lt;- &quot;levdyn&quot;
  }
  if (x_level != &quot;diverged&quot;) {
    dir_str &lt;- NULL
  }
  if (x_level == &quot;diverged&quot;) {
    if (x_effect_size &gt; 0) {
      dir_str &lt;- &quot;upcer&quot;
    }
    if (x_effect_size &lt; 0) {
      dir_str &lt;- &quot;uppar&quot;
    }
  }
  if (identical(x_cer, x_par)) {
    full_str &lt;- paste0(type_str, dir_str, x_cer)
  }
  if (!identical(x_cer, x_par)) {
    full_str &lt;- paste0(type_str, dir_str, x_cer, x_par)
  }
  return(full_str)
}) |&gt; unlist()</code></pre>
</div>
</div>
<div id="saving-and-exporting-final-data-frame-to-supplementary-table-1"
class="section level2">
<h2>Saving and exporting final data frame to Supplementary Table 1</h2>
<pre class="r"><code>save(finaldf, colordf, ExperimentNames,
     LongExperimentNames, file = &quot;data_files/FinalDataframe3Disp.RData&quot;)</code></pre>
<p>Renaming some column names and selecting specific finaldf columns for
the supplementary table, so that it appears as close to the workflow
figure as possible</p>
<pre class="r"><code>supp_tab1 &lt;- finaldf |&gt; select(gene_name, experiment, effect_size_species, padj_species, cer, par, level, plasticity) |&gt;
  dplyr::rename(c(&quot;gene&quot;=&quot;gene_name&quot;, &quot;log2_fold_change&quot;=&quot;effect_size_species&quot;,
                  &quot;padj&quot;=&quot;padj_species&quot;, &quot;cluster_Scer&quot;=&quot;cer&quot;, &quot;cluster_Spar&quot;=&quot;par&quot;))
# changing cluster names for clarity
# 0 = &quot;static&quot;
# 1 = &quot;increasing&quot;
# 2 = &quot;decreasing&quot;
supp_tab1$cluster_Scer &lt;-
  if_else(supp_tab1$cluster_Scer == 0,
          true = &quot;static&quot;,
          false = if_else(supp_tab1$cluster_Scer == 1,
                          true = &quot;increasing&quot;,
                          false = &quot;decreasing&quot;))
supp_tab1$cluster_Spar &lt;-
  if_else(supp_tab1$cluster_Spar == 0,
          true = &quot;static&quot;,
          false = if_else(supp_tab1$cluster_Spar == 1,
                          true = &quot;increasing&quot;,
                          false = &quot;decreasing&quot;))
write_delim(supp_tab1, delim = &quot;,&quot;, file = &quot;data_files/Supplementary_Table1.csv&quot;,
            col_names = TRUE)</code></pre>
</div>
<div id="gene-ontology-enrichment" class="section level2">
<h2>Gene Ontology Enrichment</h2>
<p>Creating a GOslim table recording all the gene ontology terms
associated with each gene (filtering out a few terms that are a little
too vague to be useful, such as “cellular process”).</p>
<pre class="r"><code># table of gene ontology terms associated with each gene
# each row is a term, so this is a very long table
goslim &lt;- read.table(&quot;data_files/downloaded_genomes_and_features/go_slim_mapping.tab&quot;, header = FALSE, sep = &quot;\t&quot;) |&gt;
  as_tibble()
colnames(goslim) &lt;- c(&quot;ORF&quot;, # (mandatory) - Systematic name of the gene (or gene complex if it starts with CPX-)
                      &quot;gene&quot;, # (optional) - Gene name, if one exists
                      &quot;SGDID&quot;, # SGDID (mandatory) - the SGDID, unique database identifier for the gene
                      &quot;GO_aspect&quot;, # (mandatory) - which ontology: P=Process, F=Function, C=Component
                      &quot;GOslim_term&quot;, # (mandatory) - the name of the GO term that was selected as a GO Slim term
                      &quot;GOID&quot;, # (optional) - the unique numerical identifier of the GO term
                      &quot;feature_type&quot;) # (mandatory) - a description of the sequence feature, such as ORF or tRNA
too_vague_terms &lt;- c(&quot;cellular process&quot;, &quot;molecular function&quot;, &quot;biological process&quot;)
goslim &lt;- filter(goslim, !(GOslim_term %in% too_vague_terms) &amp; 
                   (ORF %in% finaldf$gene_name))
save(goslim, file = &quot;data_files/GO_Slim.RData&quot;)</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
