<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Level and plasticity divergence in a common trans-regulatory environment</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Plasticity</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    1-3: Data Cleaning
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="clean_data.html">1. Import data and QC</a>
    </li>
    <li>
      <a href="data_for_analysis_scripts.html">2. Data cleaning</a>
    </li>
    <li>
      <a href="functions_for_figure_scripts.html">3. Functions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    4-6: Data Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="clustering.html">4. Clustering</a>
    </li>
    <li>
      <a href="single_gene_models.html">5. Generalized linear models</a>
    </li>
    <li>
      <a href="data_for_figure_scripts.html">6. Final dataframe</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    7: Single-environment figures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Fig_DiauxicShift.html">Fig 5: Diauxic Shift</a>
    </li>
    <li>
      <a href="Fig_HUShock.html">Fig S3: HU Shock</a>
    </li>
    <li>
      <a href="Fig_LowNitrogen.html">Fig S3: Low Nitrogen</a>
    </li>
    <li>
      <a href="Fig_LowPhosphate.html">Fig S3: Low Phosphate</a>
    </li>
    <li>
      <a href="Fig_HeatStress.html">Fig S3: Heat Stress</a>
    </li>
    <li>
      <a href="Fig_ColdStress.html">Fig S3: Cold Stress</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    8-10: Multi-environment figures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Fig_gene_ontology.html">8. Table 1: Gene Ontology</a>
    </li>
    <li>
      <a href="Fig_environmental_patterns.html">9. Figs 4, 6, 7: Environmental Patterns</a>
    </li>
    <li>
      <a href="Fig_cis_trans.html">10. Figs 8: F1 Hybrid</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Level and plasticity divergence in a common
trans-regulatory environment</h1>

</div>


<pre class="r"><code>source(&quot;functions_for_figure_scripts.R&quot;)
library(ggExtra)
load(&quot;data_files/FinalDataframe3Disp.RData&quot;)
load(&quot;data_files/Cleaned_Counts.RData&quot;)
load(&quot;data_files/Cleaned_Counts_Allele.RData&quot;)

cor_thresh &lt;- 0.8
eff_thresh &lt;- log2(1.5)
# extreme cor and log2 fold change values to extract the most
# cis plasticity divergers and most trans level divergers for
# avg expression figures
cis_dyn_thresh &lt;- -0.25 
trans_lev_thresh &lt;- log2(2)</code></pre>
<div id="log2-fold-changes-between-parents-and-hybrid-are-correlated"
class="section level2">
<h2>Log2 fold changes between parents and hybrid are correlated</h2>
<pre class="r"><code># R squared for effect sizes between parents and hybrid for all genes
mod_full &lt;- lm(effect_size_allele ~ effect_size_species, data = finaldf)
summary(mod_full)</code></pre>
<pre><code>## 
## Call:
## lm(formula = effect_size_allele ~ effect_size_species, data = finaldf)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -18.053  -0.288   0.002   0.289  36.082 
## 
## Coefficients:
##                      Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)         -0.085010   0.005148  -16.51   &lt;2e-16 ***
## effect_size_species  0.618360   0.004284  144.33   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 0.816 on 25254 degrees of freedom
## Multiple R-squared:  0.452,  Adjusted R-squared:  0.452 
## F-statistic: 2.083e+04 on 1 and 25254 DF,  p-value: &lt; 2.2e-16</code></pre>
<pre class="r"><code># R squared genes with consistent effect size and diverged level in at least one environment
levdiv_genes &lt;-  filter(finaldf, level == &quot;diverged&quot;) |&gt; 
  select(gene_name) |&gt; pull() |&gt; unique()
plotdf &lt;- finaldf |&gt; filter(gene_name %in% levdiv_genes) |&gt; 
  group_by(gene_name) |&gt; summarise(n_lev_dir = length(sign(effect_size_species))) |&gt; 
  filter(n_lev_dir == 1)
mod &lt;- lm(effect_size_allele ~ effect_size_species, data = filter(finaldf, gene_name %in% plotdf$gene_name))
summary(mod) # more correlated for genes with the most consistent level divergence across envrironments</code></pre>
<pre><code>## 
## Call:
## lm(formula = effect_size_allele ~ effect_size_species, data = filter(finaldf, 
##     gene_name %in% plotdf$gene_name))
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -2.7550 -0.5028  0.0341  0.4596  3.8635 
## 
## Coefficients:
##                     Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)          -0.2588     0.1270  -2.038   0.0453 *  
## effect_size_species   0.5236     0.0440  11.899   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 1.075 on 70 degrees of freedom
## Multiple R-squared:  0.6692, Adjusted R-squared:  0.6645 
## F-statistic: 141.6 on 1 and 70 DF,  p-value: &lt; 2.2e-16</code></pre>
<pre class="r"><code># looping through each experiment
library(ggExtra)
plotlist &lt;- vector(mode = &quot;list&quot;, length = length(ExperimentNames))
names(plotlist) &lt;- ExperimentNames
for (e in ExperimentNames) {
  plotdf &lt;- finaldf |&gt; filter(experiment == e &amp; level == &quot;diverged&quot;)
  mod_e &lt;- lm(effect_size_allele ~ effect_size_species, data = plotdf)
  plotlist[[e]] &lt;- ggMarginal(ggplot(plotdf, aes(x = effect_size_species,
                                 y = effect_size_allele)) +
               geom_vline(xintercept = 0, alpha = 0.5) +
               geom_hline(yintercept = 0, alpha = 0.5) +
               geom_abline(slope = 1, intercept = 0, alpha = 0.5) +
               geom_point(aes(color = experiment)) +
               geom_text(x = -2, y = 4.5, 
                         label = paste0(&quot;R^2 = &quot;, round(summary(mod_e)$r.squared,
                                                        digits = 3))) +
               scale_color_discrete(limits = colordf[colordf$scheme == &quot;experiment&quot;,]$limits,
                                    type = colordf[colordf$scheme == &quot;experiment&quot;,]$type) +
               xlim(c(-5, 5)) +
               ylim(c(-5, 5)) +
               xlab(&quot;&quot;) +
               ylab(&quot;&quot;) +
               ggtitle(LongExperimentNames[which(ExperimentNames == e)]) +
               theme_classic() +
               theme(legend.position = &quot;none&quot;),
             groupColour = TRUE, groupFill = TRUE)
}</code></pre>
<pre><code>## Warning: Removed 8 rows containing missing values or values outside the scale
## range (`geom_point()`).
## Removed 8 rows containing missing values or values outside the scale
## range (`geom_point()`).</code></pre>
<pre><code>## Warning: Removed 4 rows containing missing values or values outside the scale
## range (`geom_point()`).
## Removed 4 rows containing missing values or values outside the scale
## range (`geom_point()`).</code></pre>
<pre><code>## Warning: Removed 3 rows containing missing values or values outside the scale
## range (`geom_point()`).
## Removed 3 rows containing missing values or values outside the scale
## range (`geom_point()`).</code></pre>
<pre><code>## Warning: Removed 1 row containing missing values or values outside the scale
## range (`geom_point()`).
## Removed 1 row containing missing values or values outside the scale
## range (`geom_point()`).</code></pre>
<pre><code>## Warning: Removed 10 rows containing missing values or values outside the
## scale range (`geom_point()`).
## Removed 10 rows containing missing values or values outside the
## scale range (`geom_point()`).</code></pre>
<pre><code>## Warning: Removed 12 rows containing missing values or values outside the
## scale range (`geom_point()`).
## Removed 12 rows containing missing values or values outside the
## scale range (`geom_point()`).</code></pre>
<pre class="r"><code>pdf(&quot;paper_figures/CisTrans/l2fc_eachEnvironment.pdf&quot;,
    width = 5, height = 8)
ggarrange(plotlist = plotlist, nrow = 3, ncol = 2)
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<pre class="r"><code>ggarrange(plotlist = plotlist, nrow = 3, ncol = 2)</code></pre>
<p><img src="Fig_cis_trans_files/figure-html/l2fc-correlated-1.png" width="672" /></p>
</div>
<div
id="hybrids-have-more-correlated-plasticity-between-alleles-than-parental-orthologs"
class="section level2">
<h2>Hybrids have more correlated plasticity between alleles than
parental orthologs</h2>
<pre class="r"><code>#### Distribution of parent/hybrid expression correlation ####
plotlist &lt;- vector(mode = &quot;list&quot;, length = length(ExperimentNames))
names(plotlist) &lt;- ExperimentNames
# density plots of parental vs hybrid cor in each environment
for (e in ExperimentNames) {
  plotdf &lt;- finaldf |&gt; filter(experiment == e &amp; plasticity == &quot;diverged&quot;) |&gt; 
    select(gene_name, experiment, cor_parents, cor_hybrid) |&gt; 
    pivot_longer(cols = c(&quot;cor_parents&quot;, &quot;cor_hybrid&quot;), 
                 names_to = &quot;parent_or_hybrid&quot;,
                 values_to = &quot;cor&quot;)
  plotlist[[e]] &lt;- ggplot(plotdf, aes(x = cor)) + 
    geom_density(data = filter(plotdf, parent_or_hybrid == &quot;cor_parents&quot;), 
                 aes(fill = experiment), alpha = 1, color = &quot;white&quot;) +
    geom_density(data = filter(plotdf, parent_or_hybrid == &quot;cor_hybrid&quot;), 
                 aes(fill = experiment), alpha = 0.5, color = &quot;black&quot;) +
    scale_fill_discrete(limits = colordf[colordf$scheme == &quot;experiment&quot;,]$limits,
                        type = colordf[colordf$scheme == &quot;experiment&quot;,]$type) +
    xlim(c(-1, 1)) +
    xlab(&quot;&quot;) +
    ylab(&quot;&quot;) +
    ggtitle(LongExperimentNames[which(ExperimentNames == e)]) +
    theme_classic() +
    theme(legend.position = &quot;none&quot;)
}
p_null &lt;- ggplot(plotdf) + theme_void()
pdf(&quot;paper_figures/CisTrans/cor_eachEnvironment.pdf&quot;,
    width = 8, height = 4)
ggarrange(plotlist[[1]], plotlist[[2]],
          p_null, p_null, plotlist[[3]], plotlist[[4]],
          plotlist[[5]], plotlist[[6]], nrow = 2, ncol = 4)</code></pre>
<pre><code>## Warning: Removed 2 rows containing non-finite outside the scale range
## (`stat_density()`).</code></pre>
<pre><code>## Warning: Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).</code></pre>
<pre><code>## Warning: Removed 2 rows containing non-finite outside the scale range
## (`stat_density()`).
## Removed 2 rows containing non-finite outside the scale range
## (`stat_density()`).</code></pre>
<pre><code>## Warning: Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).</code></pre>
<pre class="r"><code>dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<pre class="r"><code>ggarrange(plotlist[[1]], plotlist[[2]],
          p_null, p_null, plotlist[[3]], plotlist[[4]],
          plotlist[[5]], plotlist[[6]], nrow = 2, ncol = 4)</code></pre>
<pre><code>## Warning: Removed 2 rows containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).</code></pre>
<pre><code>## Warning: Removed 2 rows containing non-finite outside the scale range
## (`stat_density()`).
## Removed 2 rows containing non-finite outside the scale range
## (`stat_density()`).</code></pre>
<pre><code>## Warning: Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).
## Removed 1 row containing non-finite outside the scale range
## (`stat_density()`).</code></pre>
<p><img src="Fig_cis_trans_files/figure-html/plasticity-cor-1.png" width="672" /></p>
</div>
<div id="level-divergence-tends-to-be-in-cis" class="section level2">
<h2>Level divergence tends to be in cis</h2>
<p>What better way to show this than that looking at the small fraction
of genes that are diverging in level in trans the most—that are
diverging in level in the parents and have the smallest log2 fold change
between hybrid alleles.</p>
<pre class="r"><code># pct genes in &quot;transest level&quot; group for figure
plotdf &lt;- finaldf |&gt; 
  filter(level == &quot;diverged&quot;) |&gt; 
  mutate(is_trans_upcer = (effect_size_species &gt; trans_lev_thresh) &amp;
           (abs(effect_size_allele) &lt; eff_thresh),
         is_trans_uppar = (effect_size_species &lt; -trans_lev_thresh) &amp;
           (abs(effect_size_allele) &lt; eff_thresh))
plotdf |&gt; mutate(is_trans = is_trans_upcer | is_trans_uppar) |&gt; 
  group_by(experiment) |&gt; 
  summarise(pct_trans = sum(is_trans)/n(),
            pct_rest = sum(!is_trans)/n())</code></pre>
<pre><code>## # A tibble: 6 × 3
##   experiment pct_trans pct_rest
##   &lt;chr&gt;          &lt;dbl&gt;    &lt;dbl&gt;
## 1 CC             0.173    0.827
## 2 Cold           0.221    0.779
## 3 HAP4           0.148    0.852
## 4 Heat           0.236    0.764
## 5 LowN           0.135    0.865
## 6 LowPi          0.242    0.758</code></pre>
<pre class="r"><code>pctdf &lt;- plotdf |&gt; mutate(is_trans = is_trans_upcer | is_trans_uppar) |&gt; 
  group_by(experiment) |&gt; 
  summarise(trans = sum(is_trans)/n(),
            cis = sum(!is_trans)/n()) |&gt; 
  pivot_longer(cols = c(&quot;cis&quot;, &quot;trans&quot;), names_to = &quot;mode&quot;,
               values_to = &quot;percent&quot;)
p &lt;- ggplot(pctdf, aes(x=&quot;&quot;, y=percent, fill=mode)) +
  scale_fill_discrete(limits = c(&quot;cis&quot;, &quot;trans&quot;),
                      type = c(&quot;salmon&quot;, &quot;aquamarine&quot;)) +
  geom_bar(stat=&quot;identity&quot;, width=1) +
  coord_polar(&quot;y&quot;, start=0) +
  theme_classic() +
  theme(axis.ticks = element_blank(),
        axis.text = element_blank(),
        legend.position = &quot;bottom&quot;) +
  xlab(&quot;&quot;) +
  ylab(&quot;&quot;) +
  facet_wrap(~factor(experiment, levels = ExperimentNames,
                     labels = LongExperimentNames), nrow = 2, ncol = 3)
pdf(&quot;paper_figures/CisTrans/level_pcts.pdf&quot;,
    width = 8, height = 4)
p
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<pre class="r"><code>p</code></pre>
<p><img src="Fig_cis_trans_files/figure-html/level-pie-chart-1.png" width="672" /></p>
<p>Average expression of the cis and trans portion of level divergers,
Diauxic Shift as example:</p>
<pre class="r"><code>plotdf &lt;- finaldf |&gt; 
  filter(level == &quot;diverged&quot;) |&gt; 
  mutate(is_trans_upcer = (effect_size_species &gt; trans_lev_thresh) &amp;
           (abs(effect_size_allele) &lt; eff_thresh),
         is_trans_uppar = (effect_size_species &lt; -trans_lev_thresh) &amp;
           (abs(effect_size_allele) &lt; eff_thresh))
plotdf |&gt; mutate(is_trans = is_trans_upcer | is_trans_uppar) |&gt; 
  group_by(experiment) |&gt; 
  summarise(pct_trans = sum(is_trans)/n(),
            pct_rest = sum(!is_trans)/n())</code></pre>
<pre><code>## # A tibble: 6 × 3
##   experiment pct_trans pct_rest
##   &lt;chr&gt;          &lt;dbl&gt;    &lt;dbl&gt;
## 1 CC             0.173    0.827
## 2 Cold           0.221    0.779
## 3 HAP4           0.148    0.852
## 4 Heat           0.236    0.764
## 5 LowN           0.135    0.865
## 6 LowPi          0.242    0.758</code></pre>
<pre class="r"><code># upcer, trans
gene_idxs &lt;- filter(plotdf, experiment == &quot;HAP4&quot; &amp; cer == 1 &amp; 
                      par == 1 &amp; is_trans_upcer) |&gt; 
  select(gene_name) |&gt; pull()
p &lt;- annotate_figure(plotGenes(gene_idxs, .quartet = TRUE,
                               .experiment_name = &quot;HAP4&quot;), 
                     top = paste(length(gene_idxs), &quot;genes&quot;))</code></pre>
<pre><code>## `summarise()` has grouped output by &#39;group_id&#39;, &#39;gene_name&#39;,
## &#39;experiment&#39;. You can override using the `.groups` argument.
## `summarise()` has grouped output by &#39;time_point_num&#39;, &#39;experiment&#39;.
## You can override using the `.groups` argument.
## Adding missing grouping variables: `time_point_num`, `experiment`
## Adding missing grouping variables: `time_point_num`, `experiment`</code></pre>
<pre class="r"><code>p</code></pre>
<p><img src="Fig_cis_trans_files/figure-html/level-avg-expr-1.png" width="672" /></p>
<pre class="r"><code>pdf(&quot;paper_figures/CisTrans/level_trans_HAP411_upcer.pdf&quot;, width = 3, height = 3)
p
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<pre class="r"><code># upcer, cis
gene_idxs &lt;- filter(plotdf, experiment == &quot;HAP4&quot; &amp; cer == 1 &amp; 
                      par == 1 &amp; !is_trans_upcer &amp; !is_trans_uppar &amp;
                        effect_size_species &gt; 0) |&gt; 
  select(gene_name) |&gt; pull()
p &lt;- annotate_figure(plotGenes(gene_idxs, .quartet = TRUE,
                               .experiment_name = &quot;HAP4&quot;), 
                     top = paste(length(gene_idxs), &quot;genes&quot;))</code></pre>
<pre><code>## `summarise()` has grouped output by &#39;group_id&#39;, &#39;gene_name&#39;,
## &#39;experiment&#39;. You can override using the `.groups` argument.
## `summarise()` has grouped output by &#39;time_point_num&#39;, &#39;experiment&#39;.
## You can override using the `.groups` argument.
## Adding missing grouping variables: `time_point_num`, `experiment`
## Adding missing grouping variables: `time_point_num`, `experiment`</code></pre>
<pre class="r"><code>p</code></pre>
<p><img src="Fig_cis_trans_files/figure-html/level-avg-expr-2.png" width="672" /></p>
<pre class="r"><code>pdf(&quot;paper_figures/CisTrans/level_cis_HAP411_upcer.pdf&quot;, width = 3, height = 3)
p
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
</div>
<div id="plasticity-divergence-tends-to-be-in-trans"
class="section level2">
<h2>Plasticity divergence tends to be in trans</h2>
<pre class="r"><code>plotdf &lt;- finaldf |&gt; 
  filter(plasticity == &quot;diverged&quot;) |&gt; 
  mutate(is_cis = cor_hybrid &lt; cis_dyn_thresh)
plotdf |&gt; group_by(experiment) |&gt; 
  summarise(cis = sum(is_cis, na.rm = TRUE)/n(),
            trans = sum(!is_cis, na.rm = TRUE)/n())</code></pre>
<pre><code>## # A tibble: 6 × 3
##   experiment    cis trans
##   &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt;
## 1 CC         0.161  0.838
## 2 Cold       0.103  0.896
## 3 HAP4       0.148  0.852
## 4 Heat       0.0321 0.967
## 5 LowN       0.0858 0.913
## 6 LowPi      0.105  0.893</code></pre>
<pre class="r"><code>pctdf &lt;- plotdf |&gt; group_by(experiment) |&gt; 
  summarise(cis = sum(is_cis, na.rm = TRUE)/n(),
            trans = sum(!is_cis, na.rm = TRUE)/n()) |&gt; 
  pivot_longer(cols = c(&quot;cis&quot;, &quot;trans&quot;), names_to = &quot;mode&quot;,
               values_to = &quot;percent&quot;)
p &lt;- ggplot(pctdf, aes(x=&quot;&quot;, y=percent, fill=mode)) +
  scale_fill_discrete(limits = c(&quot;cis&quot;, &quot;trans&quot;),
                      type = c(&quot;salmon&quot;, &quot;aquamarine&quot;)) +
  geom_bar(stat=&quot;identity&quot;, width=1) +
  coord_polar(&quot;y&quot;, start=0) +
  theme_classic() +
  theme(axis.ticks = element_blank(),
        axis.text = element_blank(),
        legend.position = &quot;bottom&quot;) +
  xlab(&quot;&quot;) +
  ylab(&quot;&quot;) +
  facet_wrap(~factor(experiment, levels = ExperimentNames,
                     labels = LongExperimentNames), nrow = 2, ncol = 3)
p</code></pre>
<p><img src="Fig_cis_trans_files/figure-html/plasticity-pie-chart-1.png" width="672" /></p>
<pre class="r"><code>pdf(&quot;paper_figures/CisTrans/plasticity_pcts.pdf&quot;,
    width = 8, height = 4)
p
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<p>Average expression of Diauxic Shift example:</p>
<pre class="r"><code>plotdf &lt;- finaldf |&gt; 
  filter(plasticity == &quot;diverged&quot;) |&gt; 
  mutate(is_cis = cor_hybrid &lt; cis_dyn_thresh)
plotdf |&gt; group_by(experiment) |&gt; 
  summarise(cis = sum(is_cis, na.rm = TRUE)/n(),
            trans = sum(!is_cis, na.rm = TRUE)/n())</code></pre>
<pre><code>## # A tibble: 6 × 3
##   experiment    cis trans
##   &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt;
## 1 CC         0.161  0.838
## 2 Cold       0.103  0.896
## 3 HAP4       0.148  0.852
## 4 Heat       0.0321 0.967
## 5 LowN       0.0858 0.913
## 6 LowPi      0.105  0.893</code></pre>
<pre class="r"><code># cis
gene_idxs &lt;- filter(plotdf, experiment == &quot;HAP4&quot; &amp; cer == 2 &amp; 
                        par == 1 &amp; is_cis) |&gt; 
    select(gene_name) |&gt; pull()
p &lt;- annotate_figure(plotGenes(gene_idxs, .quartet = TRUE,
                                    .experiment_name = &quot;HAP4&quot;), 
                          top = paste(length(gene_idxs), &quot;genes&quot;))</code></pre>
<pre><code>## `summarise()` has grouped output by &#39;group_id&#39;, &#39;gene_name&#39;,
## &#39;experiment&#39;. You can override using the `.groups` argument.
## `summarise()` has grouped output by &#39;time_point_num&#39;, &#39;experiment&#39;.
## You can override using the `.groups` argument.
## Adding missing grouping variables: `time_point_num`, `experiment`
## Adding missing grouping variables: `time_point_num`, `experiment`</code></pre>
<pre class="r"><code>p</code></pre>
<p><img src="Fig_cis_trans_files/figure-html/plasticity-avg-expr-1.png" width="672" /></p>
<pre class="r"><code>pdf(&quot;paper_figures/CisTrans/plasticity_cis_HAP421.pdf&quot;,
        width = 3, height = 3)
p
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<pre class="r"><code># trans
gene_idxs &lt;- filter(plotdf, experiment == &quot;HAP4&quot; &amp; cer == 2 &amp; 
                        par == 1 &amp; !is_cis) |&gt; 
    select(gene_name) |&gt; pull()
p &lt;- annotate_figure(plotGenes(gene_idxs, .quartet = TRUE,
                                    .experiment_name = &quot;HAP4&quot;), 
                          top = paste(length(gene_idxs), &quot;genes&quot;))</code></pre>
<pre><code>## `summarise()` has grouped output by &#39;group_id&#39;, &#39;gene_name&#39;,
## &#39;experiment&#39;. You can override using the `.groups` argument.
## `summarise()` has grouped output by &#39;time_point_num&#39;, &#39;experiment&#39;.
## You can override using the `.groups` argument.
## Adding missing grouping variables: `time_point_num`, `experiment`
## Adding missing grouping variables: `time_point_num`, `experiment`</code></pre>
<pre class="r"><code>p</code></pre>
<p><img src="Fig_cis_trans_files/figure-html/plasticity-avg-expr-2.png" width="672" /></p>
<pre class="r"><code>pdf(&quot;paper_figures/CisTrans/plasticity_trans_HAP421.pdf&quot;,
        width = 3, height = 3)
p
dev.off()</code></pre>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
